### DISCUSSION TITLE
Module authoring nuxt 3

### DISCUSSION THREAD

**Discussion by odranoelBR** [nuxt/nuxt#18861] [Questions]
Hello community o/

I have authored a module and i have a problem on the applications using it.
One dependency used by keycloak-js ( lib to handle auth ) have not exported 'default'.

So, i tried to make the dependency es6 with vite optimizeDeeps.include on the module, but i think it's not optimizing to final build.
Tried build transpile and it doesn't work.

When i run playground it works, but after build and release, install on a real project crash.

```js 
  async setup (options, nuxt) {
    const { resolve } = createResolver(import.meta.url)
    const logger = useLogger('Module')

    logger.info('Install module..')

    /** Add modules */
    installModule('@pinia/nuxt', { autoImports: ['defineStore', 'acceptHMRUpdate'] })
    installModule('@sidebase/nuxt-session')

    /** Add plugins */
    addPlugin(resolve('./runtime/plugins/keycloak.client'))
    addPlugin(resolve('./runtime/plugins/primevue'))
    
    /** Custom Build */
    nuxt.options.build.transpile.push('primevue')

    nuxt.options.runtimeConfig.public.apiBaseUrl = ''
    nuxt.options.runtimeConfig.public.keycloakUrl = ''
    nuxt.options.runtimeConfig.public.keycloakRealm = ''
    nuxt.options.runtimeConfig.public.keycloakClientId = ''

    nuxt.options.css.push('primevue/resources/primevue.css')
    nuxt.options.css.push('primeicons/primeicons.css')
    nuxt.options.css.push('primeflex/primeflex.css')
    nuxt.options.css.push('primevue/resources/themes/saga-blue/theme.css')

    /** Imports */
    addImportsDir(resolve('./stores'))

    /** Register components */
    await addComponentsDir({
      path: resolve('./runtime/components/layout'),
      pathPrefix: false,
      prefix: 'Layout',
      global: true
    })

    logger.success('Module success install!')
  }
})
```

**Comment by odranoelBR** [comment]
![Screenshot from 2023-02-07 10-19-07](https://user-images.githubusercontent.com/7774534/217256075-07293114-2c9f-48bc-ae7d-8f90f0a0ae8b.png)

**Comment by Kingside88** [comment]
Hi, i am also in the situation struggeling with keycloak.
My first attempt was to use sidebase nuxt-auth alongside keycloke provider.
Yes it works, but is has some bugs like SignOut Session clearing and the JWT token is not valid.
You can check out my ready to go project here: 
https://github.com/Kingside88/nuxt3-primevue-starter-auth

My second try is to implement clean vue3 with keycloakJs. This works very well without bugs.
You can see here the main branch:
https://github.com/Kingside88/vue3-auth-keycloakJs

Now I want to implement nuxt3 with keycloakJs with plugins. But get error 500 - document not defined.
As I know this issue has something to to with client/server compliance.
https://github.com/Kingside88/nuxt3-primevue-starter-auth/tree/keycloakJs

I hope this helps you and or you can provide me some help with your'e knowledge

**Reply by odranoelBR** [reply]
Tks @Kingside88 , very good info!

To solve this document is not defined, used nuxt feature to run plugins only client side. The name will be "keycloak.client.ts"
The feature https://nuxtjs.org/docs/directory-structure/plugins/#client-or-server-side-only

https://github.com/Kingside88/nuxt3-primevue-starter-auth/blob/keycloakJs/plugins/keycloak.ts

**Comment by Kingside88** [comment]
I've got an update for you. Nuxt 3 is running with official keycloakJs Javascript package.
You can get a ready to go demo here: 
https://github.com/Kingside88/nuxt3-primevue-starter-auth/blob/keycloakJs/README.md
The sidebase/nuxt-auth or better say next-auth keycloak provider has a bug providing the JWT token. It misses some properties. Now we can compare both solutions.

**Reply by odranoelBR** [reply]
Hello @Kingside88 very good man.

I solved my problem with nuxt-layers, my problem isn't related to keycloak, but about publishing nuxt modules with inside dependencies. The nuxt-layers is a better approach for what i need.

Btw this nuxt 3 is running only client side right ? i see ssr false on config. 
Use keycloak without ssr is working, but when we try to authenticate the user after the first load in the server side / node is not easy.
Lets keep sharing infos about KC, thansk!

**Comment by Kingside88** [comment]
Yes your'e right. It's client side only. The keycloak team had 2 libraries available "keycloakJs" for client side and "keycloak-connect" for servers ide. But they stopped the work on keycloak-connect. So I would no longer use it.
